src/Microsoft.Extensions.Configuration.AzureAppConfiguration/Models/KeyValueSelector.cs

```diff
@@ -23,5 +23,31 @@ public class KeyValueSelector
         /// The name of the Azure App Configuration snapshot to use when selecting key-values for the configuration provider.
         /// </summary>
         public string SnapshotName { get; set; }
+
+        /// <summary>
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns>true if the specified object is equal to the current object; otherwise, false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (obj is KeyValueSelector selector)
+            {
+                return KeyFilter == selector.KeyFilter
+                    && LabelFilter == selector.LabelFilter
+                    && SnapshotName == selector.SnapshotName;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Serves as the hash function.
+        /// </summary>
+        /// <returns>A hash code for the current object.</returns>
+        public override int GetHashCode()
+        {
+            return SnapshotName != null ? SnapshotName.GetHashCode() : KeyFilter.GetHashCode() ^ LabelFilter.GetHashCode();
```

Comment:
`LabelFilter.GetHashCode()` can throw if `LabelFilter` is null

Reply:
I also think it makes sense to check each property for null, since `GetHashCode` shouldn't necessarily need to know the valid cases.

This is sort of related, but the `KeyValueWatcher` can throw a null-ref if you pass in `null` as the key when registering a key-value for refresh. In that scenario, `Key` being null is unexpected since there shouldn't be a null key in the configuration. It seems like we should throw `ArgumentNullException` in `Register` before the `KeyValueWatcher` throws a null-ref, since it we already do this in the `Select` method.

Reply:
> It seems like we should throw `ArgumentNullException` in `Register` before the `KeyValueWatcher` throws a null-ref, since we already do this in the `Select` method.

Good catch! @amerjusupovic could you help to create a tracking issue or PR for this bug?